src/engine/AbsoluteValueConstraint.cpp

```diff
@@ -0,0 +1,454 @@
+#include "Statistics.h"
+
+AbsoluteValueConstraint::AbsoluteValueConstraint( unsigned b, unsigned f )
+    : _b( b )
+    , _f( f )
+    , _haveEliminatedVariables( false )
+{
+    setPhaseStatus( PhaseStatus::PHASE_NOT_FIXED );
+}
+
+PiecewiseLinearConstraint *AbsoluteValueConstraint::duplicateConstraint() const
+{
+    AbsoluteValueConstraint *clone = new AbsoluteValueConstraint( _b, _f );
+    *clone = *this;
+    return clone;
+}
+
+void AbsoluteValueConstraint::restoreState( const PiecewiseLinearConstraint *state )
+{
+    const AbsoluteValueConstraint *abs = dynamic_cast<const AbsoluteValueConstraint *>( state );
+    *this = *abs;
+}
+
+void AbsoluteValueConstraint::registerAsWatcher( ITableau *tableau )
+{
+    tableau->registerToWatchVariable( this, _b );
+    tableau->registerToWatchVariable( this, _f );
+}
+
+void AbsoluteValueConstraint::unregisterAsWatcher( ITableau *tableau )
+{
+    tableau->unregisterToWatchVariable( this, _b );
+    tableau->unregisterToWatchVariable( this, _f );
+}
+
+void AbsoluteValueConstraint::notifyVariableValue( unsigned variable, double value )
+{
+    _assignment[variable] = value;
+}
+
+/*
+  The variable watcher notification callbacks, about a change in a variable's value or bounds.
+  suppose A < x_b < B, C < x_f < D
+  if variable == x_b then:
+    A < 0 & B < 0 then: max{|B|, C} < x_f < min{|A|, D}
+    A < 0 & B > 0 then: max{0, C} < x_f < min{max{|A|, B}, D}
+    A > 0 & B < 0 then: ------
+    A > 0 & B > 0 then: max{A, C} < x_f < min{B, D}
+  if variable == x_f then:
+    C > 0 & D > 0 then: min{max{-D, A}, max{A, C}} < x_b < max{min{-C, A}, min{B, D}}
+*/
+
+void AbsoluteValueConstraint::notifyLowerBound( unsigned variable, double bound )
+{
+    if ( _statistics )
+        _statistics->incNumBoundNotificationsPlConstraints();
+
+    if ( _lowerBounds.exists( variable ) &&
+         !FloatUtils::gt( bound, _lowerBounds[variable] ) )
+        return;
+
+    _lowerBounds[variable] = bound;
+
+    // Check whether the phase has become fixed
+    fixPhaseIfNeeded();
+
+    // Update partner's bound
+    if ( isActive() && _constraintBoundTightener )
+    {
+        if ( variable == _b )
+        {
+            if ( bound < 0 )
+            {
+                double fUpperBound = FloatUtils::max( -bound, _upperBounds[_b] );
+                _constraintBoundTightener->registerTighterUpperBound( _f, fUpperBound );
+            }
+            else
+            {
+                // Phase is fixed, don't care about this case
+            }
+        }
+        else
+        {
+            // F's lower bound can only cause bound propagations if it
+            // fixes the phase of the constraint, so no need to bother
+        }
+
+        // Also, if for some reason we only know a negative lower bound for f,
+        // we attempt to tighten it to 0
+        if ( bound < 0 && variable == _f )
+            _constraintBoundTightener->registerTighterLowerBound( _f, 0 );
```

Comment:
This should be done only once, ideally at the start of solving. What is the reason to keep attempting it?

Reply:
Indeed, this should only apply in the beginning, but we don't know whether it's been done before or not. I refactored it a bit, for efficiency (moved it into the previous if condition).